src/service/history-impl.js

```diff
@@ -643,15 +628,19 @@ export class HistoryBindingNatural_ {
   replaceStateForTarget(target) {
     dev().assert(target[0] == '#', 'target should start with a #');
     this.whenReady_(() => {
-      // location.replace will fire a popstate event, this is not a history
-      // event. This tells the popstate handler to not handle it by setting
-      // the lastNavigatedHash_ to the future hash we know we're going toward.
+      // location.replace will fire a popstate event which is not a history
+      // event, so temporarily remove the event listener and re-add it after.
       // As explained above in the function comment, typically we'd just do
       // replaceState here but in order to trigger :target re-eval we have to
       // use location.replace.
-      this.lastNavigatedHash_ = target;
-      // TODO(mkhatib, #6095): Chrome iOS will add extra states for location.replace.
-      this.win.location.replace(target);
+      this.win.removeEventListener('popstate', this.popstateHandler_);
```

Comment:
I'm pretty sure I suggested just using a boolean before, which'll be a bit faster.